### PR TITLE
Compute the min heap size required to a Java program.

### PR DESCRIPTION
**Pull request by ansalond**
I added a command that computes the minimum heap size required to run a Java program within a certain overhead factor.

This required some changes to `run_java` (run with a timeout) and to `_waitWithTimeout` (the timeout of a subprocess should not be fatal).

### CODE REVIEW

File: mx.py [hunk 1]
```diff
@@ -7089,7 +7126,14 @@ def _returncode(status):
             return _returncode(status)
         remaining = end - time.time()
         if remaining <= 0:
-            abort('Process timed out after {0} seconds: {1}'.format(timeout, ' '.join(args)))
+            msg = 'Process timed out after {0} seconds: {1}'.format(timeout, ' '.join(args))
+            if nonZeroIsFatal:
+                abort(msg)
+            else:
+                log(msg)
+                for p, _ in _currentSubprocesses:
```
Comment (gilles-duboscq) [inline mx.py]: Why kill all the sub-processes? why not just `process`?

File: mx.py [hunk 2]
```diff
@@ -7053,6 +7053,43 @@ def run_java(args, nonZeroIsFatal=True, out=None, err=None, cwd=None, timeout=No
         jdk = get_jdk()
     return jdk.run_java(args, nonZeroIsFatal=nonZeroIsFatal, out=out, err=err, cwd=cwd, timeout=timeout, env=env, addDefaultArgs=addDefaultArgs)
 
+def run_java_min_heap(args, benchName='# MinHeap:', overheadFactor=1.5, referenceHeap=4096, out=None, err=None, cwd=None, timeout=None, env=None, addDefaultArgs=True, jdk=None):
+    """computes the minimum heap size required to run a Java program within a certain overhead factor"""
+
+    def run_with_heap(heap, args, timeout=timeout, suppressStderr=True, nonZeroIsFatal=False):
+        log('Trying with ' + str(heap) + 'MB of heap...')
+        with open(os.devnull, 'w') as fnull:
+            exitCode = run_java(['-Xmx' + str(heap) + 'M'] + args, nonZeroIsFatal=nonZeroIsFatal, out=out, err=fnull if suppressStderr else err, cwd=cwd, timeout=timeout, env=env, addDefaultArgs=addDefaultArgs)
```
Comment (gilles-duboscq) [inline mx.py]: I think the last `-Xmx` arg is taken into account, maybe you could try to insert as the last JVM arg? I think only the `-cp` and `--classpath` JVM arg consume the next argument.

File: mx.py [hunk 3]
```diff
@@ -7089,7 +7126,14 @@ def _returncode(status):
...
         if remaining <= 0:
-            abort('Process timed out after {0} seconds: {1}'.format(timeout, ' '.join(args)))
+            msg = 'Process timed out after {0} seconds: {1}'.format(timeout, ' '.join(args))
+            if nonZeroIsFatal:
+                abort(msg)
+            else:
+                log(msg)
+                for p, _ in _currentSubprocesses:
+                    _kill_process_group(p.pid, signal.SIGKILL)
+                return 1
```
Comment (gilles-duboscq) [inline mx.py]: Maybe you can return something which is not a valid return code so it's possible to distinguish failed runs from timeouts? (e.g., `ERROR_TIMEOUT = 0x700000000 # not 32 bits`)

File: mx.py [hunk 4]
```diff
@@ -7053,6 +7053,43 @@ def run_java(args, nonZeroIsFatal=True, out=None, err=None, cwd=None, timeout=No
...
+            return 1
+        referenceTime = time.time() - t
+        maxTime = referenceTime * overheadFactor
+        log('Reference time = ' + str(referenceTime))
+        log('Maximum time = ' + str(maxTime))
+    else:
+        maxTime = None
+
+    heap = 10
+    while run_with_heap(heap, args, timeout=maxTime):
```
Comment (gilles-duboscq) [inline mx.py]: if you do the `ERROR_TIMEOUT` stuff below then maybe you should check if we get `ERROR_TIMEOUT` here and abort if you get something else which is not `0`.

File: mx.py (hunk 1 above)
Comment (ansalond) [inline mx.py]: `abort` would kill all sub-processes. I tried to keep the semantics as similar as possible, so that the only difference is that the current process survives. However, I see your point and I am going to change this in my next push.

File: mx.py (hunk 2 above)
Comment (ansalond) [inline mx.py]: Good point. I will call `extract_VM_args` and append `-Xmx` to that list.

File: mx.py (hunk 4 above)
Comment (ansalond) [inline mx.py]: This loop increases the heap size to be used in the next attempt every time that the process does not return `0`. I don't think it should make any difference if the process timed out or it failed differently (e.g., `OutOfMemoryError` or `Excessive GC Time`).

File: mx.py (hunk 3 above)
Comment (ansalond) [inline mx.py]: You are right. I don't think it matters in my use case but it might be important to others.

File: mx.py (hunk 4 above)
Comment (gilles-duboscq) [inline mx.py]: Right

File: mx.py (hunk 2 above)
Comment (gilles-duboscq) [inline mx.py]: Do you want to do that one in a separate pull request?

File: mx.py (hunk 4 above)
Comment (Peter-B-Kessler) [inline mx.py]: Do you want to start with some reasonable heap size and then binary search your way down to near the minimum, rather than (as I read it) starting with a heap size that's too small (10MB?) and growing the heap by a fixed amount (10MB then 1MB?) after each failure?  How many iterations does it typically take you to find the minimum?  How does that number of iterations compare with log2(referenceHeap)?  How long does each iteration take for benchmarks you care about?

Comment (ansalond) [inline mx.py]: I took this part from a script that we use in a different context and I simply translated it to python. In my use cases the heap sizes that I am measuring range from 5MB to 20MB, and the whole measuring process takes at most 4 mins, so the current algorithm is not so bad. However, I see your point and I will improve the algorithm.
